// react-placeholder-typing/dist/index.modern.js
import React, { useState, useEffect } from 'react';
import styled, { keyframes } from 'styled-components';

function _extends() {
  _extends = Object.assign ? Object.assign.bind() : function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];
      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }
    return target;
  };
  return _extends.apply(this, arguments);
}
function _taggedTemplateLiteralLoose(strings, raw) {
  if (!raw) {
    raw = strings.slice(0);
  }
  strings.raw = raw;
  return strings;
}

var _templateObject, _templateObject2, _templateObject3, _templateObject4;
var typing = keyframes(_templateObject || (_templateObject = _taggedTemplateLiteralLoose(["\n  from { width: 0 }\n"])));
var blinkCursor = keyframes(_templateObject2 || (_templateObject2 = _taggedTemplateLiteralLoose(["\n  0% {border-right-color: rgba(255,255,255,.75);}\n  50% {border-right-color: transparent;}\n  100% {border-right-color: rgba(255,255,255,.75);}\n"])));
var InputWrapper = styled.div(_templateObject3 || (_templateObject3 = _taggedTemplateLiteralLoose(["\n  display: flex;\n  align-items: center;\n  color: #f5f5f5;\n  border: 1px solid #f5f5f5;\n  border-radius: 8px;\n  padding: 10px;\n  width: 100%;\n"])));
var StyledInput = styled.input(_templateObject4 || (_templateObject4 = _taggedTemplateLiteralLoose(["\n  color: #f5f5f5;\n  background: none;\n  border: none;\n  padding: 0 10px;\n  width: 80%;\n  outline: none;\n  font-family: ", ";\n  font-size: 20px;\n  &::placeholder {\n    animation: ", " 2s steps(30, end), ", " .75s step-end infinite;\n    overflow: hidden;\n    white-space: nowrap;\n    border-right: .1em solid;\n  }\n"])), function (props) {
  return props.fontFamily || 'monospace';
}, typing, blinkCursor);
var ReactTypingAnimation = function ReactTypingAnimation(_ref) {
  var _ref$placeholders = _ref.placeholders,
    placeholders = _ref$placeholders === void 0 ? [] : _ref$placeholders,
    _ref$value = _ref.value,
    value = _ref$value === void 0 ? '' : _ref$value,
    _onChange = _ref.onChange,
    renderIcon = _ref.renderIcon,
    fontFamily = _ref.fontFamily,
    _ref$containerStyle = _ref.containerStyle,
    containerStyle = _ref$containerStyle === void 0 ? {} : _ref$containerStyle,
    _ref$inputStyle = _ref.inputStyle,
    inputStyle = _ref$inputStyle === void 0 ? {} : _ref$inputStyle;
  var _useState = useState(0),
    index = _useState[0],
    setIndex = _useState[1];
  var _useState2 = useState(''),
    placeholder = _useState2[0],
    setPlaceholder = _useState2[1];
  useEffect(function () {
    var place = '';
    var typingInterval = setInterval(function () {
      if (place.length < placeholders[index].length) {
        place += placeholders[index][place.length];
        setPlaceholder(place);
      } else {
        clearInterval(typingInterval);
        var deletingInterval = setInterval(function () {
          if (place.length > 0) {
            place = place.slice(0, place.length - 1);
            setPlaceholder(place);
          } else {
            clearInterval(deletingInterval);
            setIndex((index + 1) % placeholders.length);
          }
        }, 100);
      }
    }, 100);
    return function () {
      return clearInterval(typingInterval);
    };
  }, [index]);
  return /*#__PURE__*/React.createElement(InputWrapper, {
    style: _extends({}, containerStyle)
  }, renderIcon && renderIcon(), /*#__PURE__*/React.createElement(StyledInput, {
    fontFamily: fontFamily,
    placeholder: placeholder,
    onChange: function onChange(e) {
      _onChange && _onChange(e.target.value);
    },
    value: value,
    style: _extends({}, inputStyle)
  }));
};

export default ReactTypingAnimation;
